### ISSUE TITLE
(Solved) Looking for an "Extensions Expert"

### ISSUE THREAD

**Issue by Marc0ai** [lllyasviel/stable-diffusion-webui-forge#1792]
Forgive me, it's not a real issue, but I didn't know where else to post other than here or reddit....

In short, I've finished writing a script that can be used on Forge (or automatic1111 if you prefer) which works, and it works quite well.

The problem is that so far, the only thing I've managed to do is implement it as a script (you know, the ones you use at the bottom of the WebUI from the dropdown menu), but I would like to "convert" it into a proper extension, so that it appears in the "Main" tab. Despite my attempts, I haven't been able to figure it out, or rather, I don't know how to do it practically. I've only tried emulating other extensions, but without success...

So, is there anyone who could take a look?  https://pastebin.com/z3jjxw4x

Feel free to try it out; the script itself works and is used to apply photographic filters to the generations.

**Comment by DenOfEquity** [comment]
First:
```
    def show(self, is_img2img):
        return scripts.AlwaysVisible
```
Second:
`run()` is for scripts, for extensions you have many options. Check `modules\scripts.py`.
`postprocess_image()` seems ideal for you: `    def postprocess_image(self, p, pp, *script_args):`
It'll be called for each image, with the image in `pp.image`.
Third:
wrap your UI up into a gradio Accordion.

**Comment by Marc0ai** [comment]
Wow, thanks to your answer I was finally able to figure it out....
I was just using the wrong approach, thanks, I really appreciate it.
The next challenge now will be to be able to insert it into the extensions index, because I noticed that the index has not been updated for a month now, so I don't know if there is still interest from the author of A1111....
If anyone still wants to try it, the extension can be completely used and installed from here: https://github.com/Marc0ai/sd-webui-photo-refiner

**Comment by nitinmukesh** [comment]
@Marc0ai 

Is there any dependency to be installed like dlib?

**Comment by Marc0ai** [comment]
Yep, to use "Face Enhancer" you need dlib, in theory it should install itself on the first startup

**Comment by Juqowel** [comment]
> Yep, to use "Face Enhancer" you need dlib, in theory it should install itself on the first startup

It's not.

![err](https://github.com/user-attachments/assets/36b7c338-2c26-4ea9-97b3-6069fd918395)

**Comment by Marc0ai** [comment]
Damn my bad... Evidently there are problems with that installation method... I temporarily added a separate installer, now it should add dlib correctly, let me know

**Comment by Juqowel** [comment]
Installation will be a bit complicated.

![cm](https://github.com/user-attachments/assets/04bb827c-75f6-422a-9aaf-4e2e5991e44f)

**Comment by Marc0ai** [comment]
It should actually be easier than expected, I updated everything and now it should handle installing cmake too, you can try updating the extension and try again

**Comment by Marc0ai** [comment]
I literally started getting interested in python and how to implement it in the webui 3 days ago, so I was expecting this kind of problem, with a little patience I hope to solve everything 👌

**Comment by Juqowel** [comment]
> It should actually be easier than expected, I updated everything and now it should handle installing cmake too, you can try updating the extension and try again

![cm3](https://github.com/user-attachments/assets/07196596-5d37-4029-b4e1-4d16e8114339)

You should test it on a new clean system to find the correct way.

**Comment by Marc0ai** [comment]
I'm glad the extension is useful and I really appreciate the feedback. I've made some improvements to the UI, including a link to the Git page where I've added visible examples for each effect. I'll add descriptions for the effects later. Unfortunately, I'm still trying to figure out the best way to install the dlib library, which should be quite simple in theory, but it seems to cause issues for some users. If anyone knows more about this, I'd greatly appreciate it. Oh, and I've also forwarded it to the A1111 index, hoping it's still active—I haven't seen any updates in a month...
